### ISSUE TITLE
salt-cloud grains inheritance from provider to profile (2018 vs 2016)

### ISSUE THREAD

**Issue by defanator** [saltstack/salt#49226]
### Description of Issue/Question

After upgrading salt-cloud from 2016.11.6 to 2018.3.2 we've been observing a situation when provider-specific grains have not been applied to the target instance.

### Setup

Excerpt from our initial configuration that was working fine with 2016.11.6:

```
prov1:
  driver: ec2
  minion:
    master: x.x.x.x
  grains:
    prov1_grain1: foo
    prov2_grain2: bar

profile1:
  provider: prov1
  grains:
    env: nonprod
    roles:
      - role1
      - role2
      - role2
```

Target instance launched with 2016.11.6 contained all the grains merged in a proper way, i.e.:

```
prov1_grain1: foo
prov1_grain2: bar
env: nonprod
roles:
- role1
- role2
- role3
```

Target instance launched with 2018.3.2 contained just the grains from profile:

```
env: nonprod
roles:
- role1
- role2
- role3
```

We finally managed to get it working by changing provider configuration to the following:

```
prov1:
  driver: ec2
  minion:
    master: x.x.x.x
    grains:
      prov1_grain1: foo
      prov2_grain2: bar
```

I.e., `grains` tree should be under `minion` now.

However, documentation examples still contain older (legacy?) syntax:

```
# Note: This example is for /etc/salt/cloud.providers or any file in the
# /etc/salt/cloud.providers.d/ directory.

my-ec2-southeast-public-ips:
  # Set up the location of the salt master
  #
  minion:
    master: saltmaster.example.com

  # Set up grains information, which will be common for all nodes
  # using this provider
  grains:
    node_type: broker
    release: 1.0.1
```

(from https://docs.saltstack.com/en/latest/topics/cloud/aws.html)

My questions so far:
 - is current behavior correct? (grains should be nested from minion in provider config, but not in profile config),
 - should a documentation be changed accordingly?

Thanks!

**Comment by gtmanfred** [comment]
Can you test the 2018.3.3 branch?

https://github.com/saltstack/salt/pull/47665

This pr should fix this issue, and will be released in the 2018.3.3 and 2017.7.8 versions.

I believe this issue is a duplicate of https://github.com/saltstack/salt/issues/45125

**Comment by gtmanfred** [comment]
You said above

> Target instance launched with 2018.3.2 contained just the grains from profile:

So, you have launched instances with 2018.3.3? and this still happens? can you make sure that all .pyc files have been cleared out?

**Comment by defanator** [comment]
@gtmanfred I searched for similar issues, found #45125 and realized that packaged salt-common 2018.3.2 does not contain changes from #47655, so I switched to git repository for tests and picked 2018.3.3 branch.

I'll double check for stale .pyc a bit later and let you know the results.

**Comment by defanator** [comment]
@gtmanfred I've just removed all .pyc files within salt directory, returned provider configuration to the initial format:

```
prov1:
  driver: ec2
  minion:
    master: x.x.x.x
  grains:
    prov1_grain1: foo
    prov2_grain2: bar
```

and checked with launching new instance from that profile - provider's grains are missing.

Let me know if I can provide anything else to help with investigating this one.

**Comment by gtmanfred** [comment]
Ok, i can replicate this, but i need to have more time to fix it, i am marking it as a blocker for the 2018.3.4 and 2017.7.9 releases.

**Comment by sagetherage** [comment]
@gtmanfred un-assigning you and looking for a replacement, today.

**Comment by defanator** [comment]
@sagetherage @dwoz any chances to get this one fixed in 2018.3.4? TIA!

**Comment by sagetherage** [comment]
@defanator yes, our process is to label the ticket with the releases and this is also labeled with `Blocker` indicating we are working on a fix for those releases and not releasing until fixed. If anything changes you will see comments added to this ticket. I hope this is helpful information.
